dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ReplaceZeroRegisters.java

```diff
@@ -0,0 +1,50 @@
+package com.dat3m.dartagnan.program.processing;
+
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.program.Function;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.EventFactory;
+import com.dat3m.dartagnan.program.event.RegReader;
+import com.dat3m.dartagnan.program.event.RegWriter;
+import com.dat3m.dartagnan.program.event.core.Local;
+
+/*
+    This pass replaces every RegWritter
+        zr <- e
+    targeting a zero register to
+        dummy <- e; zr <- 0
+    and for every RegReader reading zr, it inserts an event zr <- 0 before it
+*/
```

Comment:
(1) Typo
(2) Since dependencies are not preserved, just replace all read-usages of `zr` directly by `0`. Alternatively, if you really want to keep the register and let SCCP do the replacement, just set it once to `0` at the beginning of the thread/function. You guarantee that the register is never overwritten, so no need to reset it all the time.

Reply:
Again, I don't want passes to do more magic than they should (thus I don't want to replace all read-usages by zero). I know CP will eventually do this, but
1) that is the purpose of CP, thus there is no magic here.
2) ATM we are not even applying CP to litmus tests (we only have zero register for litmus code), thus the local assignment is explicitly visible. I don't think we gain much performance by removing those local events, so I don't see the point.

In principle, having a single assignment at the beginning will not work because nothing prevents us of set two different registers as zero one (this was the whole purpose of your proposal of moving the `replaceZeroRegister` (now `setZeroRegister`) to the program builder). I could have code to get all different zero registers and initialize each of them once, but this will probably result in more complicated code with almost null benefit in terms of performance.

Reply:
> Again, I don't want passes to do more magic than they should (thus I don't want to replace all read-usages by zero). I know CP will eventually do this, but
> 
>     1. that is the purpose of CP, thus there is no magic here.
> 
>     2. ATM we are not even applying CP to litmus tests (we only have zero register for litmus code), thus the local assignment is explicitly visible. I don't think we gain much performance by removing those local events, so I don't see the point.

I don't understand how it is more magic to just set the registers to zero once and for all. Isn't this the more canonical solution? I don't care about SCCP or any optimization/performance. In this case, whether the register is set again and again does not really matter, but it is a more convoluted solution with a few more disadvantages (requires more code, bloats up printing or even witness generation in the future).

I think it is a lot more magic to have a "isZero" flag that must be set in the parser for some pass to come along and enforce zeroness exactly once. Afterwards, the flag is useless and actually wrong, because we never again enforce that the register holds only zero values: I can create "zr <- 1" events and no one will complain that I write non-zero values to zero registers.
And this is IMO the biggest problem of this approach: you introduce state into the core logic (`Register` class) that far outlives its single use-case and thus opens up the possiblity of inconsistency (cause zero-ness is never enforced again).

> In principle, having a single assignment at the beginning will not work because nothing prevents us of set two different registers as zero one (this was the whole purpose of your proposal of moving the `replaceZeroRegister` (now `setZeroRegister`) to the program builder). I could have code to get all different zero registers and initialize each of them once, but this will probably result in more complicated code with almost null benefit in terms of performance.

I don't understand. What stops you from initializating every zero register to 0? Just pass multiple register names to the pass on creation, or just let the pass run multiple pass. In fact, the reason I say this is because it results in overall less code:
```Java
foreach (String zeroRegName : zeroRegNames) {
   Register zeroReg = func.getRegisterByName(zeroRegName);
   if (zeroReg != null) {
        Event initToZero = EventFactory.newLocal(zeroReg, expressions.getGeneralZero(zeroreg.getType()))
        func.getEntry().insertAfter(initToZero);
   }
}
```
This piece of code subsumes your doubly-nested RegReader loop while being more clear and concise (and no nested loops!). And it allows you to get rid of all the modifications to `Register` and elsewhere, easily saving 10+ lines of code.